c8d: make the cache in classic builder work

rumpl opened the pull request:
- addresses https://github.com/moby/moby/issues/46760

**- What I did**

Makes the cache in classic builder work with the containerd image store

In order for the cache in the classic builder to work we need to:
- use the came comparison function as the graph drivers implementation
- save the container config when commiting the image
- use all images to search a 'FROM "scratch"' image
- use all images if `cacheFrom` is empty
- use the image labels to find the parents/children, not the rootfs

**- How I did it**

**- How to verify it**

```console
$ make DOCKER_GRAPHDRIVER=overlayfs TEST_FILTER=TestBuildAddCurrentDirWithoutCache TEST_INTEGRATION_USE_SNAPSHOTTER=1 TEST_IGNORE_CGROUP_CHECK=1 test-integration
```

```console
$ make DOCKER_GRAPHDRIVER=overlayfs TEST_FILTER=TestBuildAddRemoteFileWithAndWithoutCache TEST_INTEGRATION_USE_SNAPSHOTTER=1 TEST_IGNORE_CGROUP_CHECK=1 test-integration
```

```console
$ make DOCKER_GRAPHDRIVER=overlayfs TEST_FILTER=TestBuildCopyDirButNotFile TEST_INTEGRATION_USE_SNAPSHOTTER=1 TEST_IGNORE_CGROUP_CHECK=1 test-integration 
```

And many others.

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

rumpl committed:
c8d: make the cache in classic builder work

In order for the cache in the classic builder to work we need to:
- use the came comparison function as the graph drivers implementation
- save the container config when commiting the image
- use all images to search a 'FROM "scratch"' image
- load all images if `cacheFrom` is empty

Signed-off-by: Djordje Lukic <djordje.lukic@docker.com>
Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

rumpl commented:
Converted to draft, just realized that the scratch cache case needs some more work

vvoland commented on daemon/containerd/cache.go:
```diff
@@ -14,80 +15,230 @@ import (
+	// FROM scratch
 	if parentID == "" {
-		// TODO handle "parentless" image cache lookups ("FROM scratch")
-		return "", nil
+		imgs, err := ic.imageService.Images(ctx, types.ImageListOptions{
+			All: true,
+		})
+		if err != nil {
+			return "", err
+		}
+		for _, img := range imgs {
+			children = append(children, image.ID(img.ID))
+		}
+	} else {
```
I think this is incorrect. Any image is not a suitable cache for the `FROM scratch` image. This needs to be only the intermediate images that were built `FROM scratch`.

Not sure what's the best way to detect them though - because it looks like containerd doesn't store labels with empty values, so we can't really just check if `﻿org.mobyproject.image.parent` label exists but is empty.

rumpl replied:
In the original case, if it's a `FROM scratch` we take the whole image map https://github.com/moby/moby/blob/master/image/cache/cache.go#L239

vvoland replied:
But they're also filtered with `img.Parent == ""`: https://github.com/moby/moby/blob/777e9f271095685543f30df0ff7a12397676f938/image/cache/cache.go#L243

rumpl replied:
Right, let me add that, see if anything breaks

vvoland replied:
I don't think it's enough - the Parent will be empty for pulled/buildkit images so they will be considered a valid cache too.

rumpl commented on daemon/containerd/cache.go:
```diff
@@ -14,80 +15,232 @@ import (
+			}
+			for _, hi := range h {
+				if hi.ID != "<missing>" {
+					im, err := i.GetImage(ctx, hi.ID, imagetype.GetImageOpts{})
+					if err != nil {
+						return nil, err
+					}
```
I prefer to check only the images we should have locally rather than having to call back containerd when I know the image can't exist

laurazard replied:
Oh 🤦‍♀️ yeah of course, that makes sense

laurazard commented on daemon/containerd/image_children.go:
```diff
@@ -3,216 +3,56 @@ package containerd
-				continue
-			}
-
-			for _, parentRoot := range parentRootFS {
-				if isRootfsChildOf(rootfs, parentRoot) {
-					children = append(children, image.ID(img.Target.Digest))
-					break nextImage
-				}
-			}
-		}
+		children = append(children, image.ID(img.Target.Digest))
 	}
 
 	return children, nil
 }
```
We're changing the contract for this, right? (this is why we're changing the godoc, or just to remove no longer accurate description of the implementation detail?).

Is this a safe change – return all (even indirect) children of this image -> only return direct descendants? Are all the callers expecting this?

If so, then we should also make this "contract" change clearer in the godoc.

(please disregard this comment if I got the changes wrong 😅)

laurazard replied:
thinking especially re: the image list and image delete scenario

vvoland replied:
Yeah, so originally this was supposed to actually only return the direct children (at least this is what the graphdriver implementation does). The usage in `daemon/list.go` only expects the direct children (but it recurses anyway so it's fine either way).

The image delete scenario needs to be checked though.

rumpl replied:
Delete also calls `Children` and doesn't need ancestors, only direct children

vvoland committed:
c8d/cache: Optimize `FROM scratch` case

Consider only images that were built `FROM scratch` as valid candidates
for the `FROM scratch` + INSTRUCTION build step.

The images are marked as `FROM scratch` based by the classic builder
with a special label. It must be a new label instead of empty parent
label, because empty label values are not persisted.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/builder: Store ContainerConfig

Serialize ContainerConfig to content store and store its digest in
label.

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland committed:
c8d/cache: Use ContainerConfig from content store

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented:
FYI, the validate failure is caused by splitting one integration-cli test into two separate ones.

thaJeztah commented:
@vvoland oh! I saw that, but looks like I didn't comment; could you perhaps push a temporary commit similar to https://github.com/moby/moby/commit/7ed823ead91938e5cc8de7a42ff93a25abe73e7e, so that we have the validate steps run?

We can revert it after this is merged (we should have some option to allow skipping it in CI without having to make changes, but don't know what the best approach for that is with GitHub actions 😞)

vvoland committed:
temporary: Disable deprecate-integration-cli validation

Signed-off-by: Paweł Gronowski <pawel.gronowski@docker.com>

vvoland commented:
Done

laurazard commented on daemon/containerd/image_history.go:
```diff
@@ -161,3 +163,24 @@ func getImageTags(ctx context.Context, imgs []images.Image) []string {
 	return tags
 }
+
+// getParentsByBuilderLabel finds images that were a base for the given image
```
This godoc uses the plural (e.g. "finds image**s**") and we do return the result of `ImageService().List()`, but in general this should result in "one" image right? (the direct parent)

vvoland replied:
It will be one image in terms of content, but the actual tag might be different (`busybox:latest`, `busybox:3.16`, etc).

It's not new in this PR though, it was just moved to a different file

laurazard commented on daemon/containerd/image_children.go:
```diff
@@ -30,136 +26,38 @@ func (i *ImageService) Children(ctx context.Context, id image.ID) ([]image.ID, e
-	}
+	var imgs []containerdimages.Image
+	for {
+		parent, ok := targetImage.Labels[imageLabelClassicBuilderParent]
```
This looks pretty similar to `parents()` in `image_history.go`, I wonder if we can refactor them somehow?

vvoland replied:
Hmm yeah, we should be able to use getParentsByBuilderLabel here

laurazard replied:
sounds good to me!

laurazard commented on daemon/containerd/image_builder.go:
```diff
@@ -502,10 +527,17 @@ func (i *ImageService) createImageOCI(ctx context.Context, imgToCreate imagespec
+	newConfig imagespec.DockerOCIImage, layers []ocispec.Descriptor,
+	containerConfig container.Config,
+) (
+	manifestDesc ocispec.Descriptor,
+	containerConfigDesc ocispec.Descriptor,
+	_ error,
```
Nice! Reads better than just having the function return `(ocispec.Descriptor, ocispec.Descriptor, error)`

laurazard reviewed (approved):
LGTM 🎉  (yaaaaaay)

Left a couple of nitpicky comments/questions but there's probably reason for things to be that way/they're not anywhere big enough to be a blocker.

thaJeztah commented on daemon/containerd/imagespec.go:
```diff
@@ -109,7 +109,6 @@ func dockerOCIImageConfigToContainerConfig(cfg imagespec.DockerOCIImageConfig) *
 	for k, v := range cfg.ExposedPorts {
 		exposedPorts[nat.Port(k)] = v
 	}
-
```
If you're gonna update this PR (for other review comments) can you remove this change from the first commit, so that this file disappears from the final diff? 🤗

laurazard replied:
oh yes please! I didn't leave a comment because it was such a small thing and wouldn't want you to have to rebase to address just that, but I'd be happy if it happened 😅

vvoland replied:
Done!

laurazard reviewed (approved):
Re-LGTM'ing ✅

thaJeztah commented:
IT'S ALL GREEN (on Linux). IT'S ALL GREEN!! 🥳 🥳 🥳 🥳 

<img width="306" alt="Screenshot 2024-01-17 at 15 47 36" src="https://github.com/moby/moby/assets/1804568/ebfb01b7-1a5f-4551-8154-faa1e27d576e">

thaJeztah reviewed (approved):
LGTM 🥳
